docs/site/Loopback-component-authentication.md

```diff
@@ -0,0 +1,600 @@
+    private metadata?: AuthenticationMetadata,
+  ) {}
+  async value(): Promise<AuthenticationStrategy | undefined> {
+    if (!this.metadata) {
+      return undefined;
+    }
+    const name = this.metadata.strategy;
+    const strategy = await this.findAuthenticationStrategy(name);
+    if (!strategy) {
+      // important to throw a non-protocol-specific error here
+      let error = new Error(`The strategy '${name}' is not available.`);
+      Object.assign(error, {
+        code: AUTHENTICATION_STRATEGY_NOT_FOUND,
+      });
+      throw error;
+    }
+    return strategy;
+  }
+
+  async findAuthenticationStrategy(name: string) {
+    const strategies = await this.authenticationStrategies();
+    const matchingAuthStrategy = strategies.find(a => a.name === name);
+    return matchingAuthStrategy;
+  }
+}
+```
+
+{% include note.html content="
+The specific <b>name</b> that it looks for is a unique name you specify as a parameter in an <b>authentication decorator</b> applied to a controller endpoint. This decorator is discussed in the next section.
+" %}
+
+In order for your custom authentication strategy to be **found**, it needs to be
+**registered**.
+
+**Registering** a custom authentication strategy `BasicAuthenticationStrategy`
+as an extension of the
+`AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME` extension
+point in an application `application.ts` is as simple as:
+
+```ts
+registerAuthenticationStrategy(this, BasicAuthenticationStrategy);
+```
+
+## Authentication Decorator
+
+Once you have a **registered** authentication strategy with a **unique** name,
+then you are ready to use the
+[Authentication Decorator](https://loopback.io/doc/en/lb4/Decorators_authenticate.html)
+on your controller endpoints.
+
+They decorator's syntax is :
+
+```ts
+@authenticate(strategyName: string, options?: Object)
+```
+
+The strategy `name` is **mandatory** and the `options` object is **optional**.
+When the options object is specified, it must be relevant to that particular
+strategy.
+
+To **secure** a specific endpoint, you simply need to add the decorator and
+specify the `name` of a registered authentication strategy.
+
+Here is an example of the decorator using the 'basic' authentication strategy we
+defined earlier, **without** options, for the endpoint `/whoami` in a controller
+named `WhoAmIController`.
+
+```ts
+import {inject} from '@loopback/context';
+import {
+  AuthenticationBindings,
+  UserProfile,
+  authenticate,
+} from '@loopback/authentication';
+import {get} from '@loopback/rest';
+
+export class WhoAmIController {
+  constructor(
+    @inject(AuthenticationBindings.CURRENT_USER)
+    private userProfile: UserProfile,
+  ) {}
+
+  @authenticate('basic')
+  @get('/whoami')
+  whoAmI(): string {
+    return this.userProfile.id;
+  }
+}
+```
+
+An example of the decorator when options **are** specified looks like this:
+
+```ts
+@authenticate('basic', { propertyOne : "foo", propertyTwo:"bar"})
```

Comment:
nitpick: do you have some more realistic values that our users might want to use the options for?

Reply:
@dhmlau 

> nitpick: do you have some more realistic values that our users might want to use the options for?

Not really because we don't ship any authentication strategies at all ( just the interface). I could put a made up options object `{ useCookie: true}` or `{session:true}` or `{ enforceLongPassword: true}` but this would give them the wrong impression that our interface supports this kind of option.  @jannyHou @raymondfeng , what are your thoughts on this?

Reply:
I was also searching for some meaningful options for the basic auth in `passport-http`, but didn't find any...like what Dom said, they are all related to the `passport` middleware instead of the strategy itself.

Reply:
got it. maybe at least mention those options are specific to the strategy the user select?

Reply:
@jannyHou 

> I was also searching for some meaningful options for the basic auth in passport-http, but unfortunately didn't find any...like what Dom said, they are all related to the passport middleware instead of the strategy itself. 

Actually, the focus of the discussion is not on **passport** stuff in this discussion. The `@authenticate(strategyName, options?)` decorator allows for passing in `options`. This gives the developer of custom authentication strategies a way to pass options in...if needed.

(Buy Janny is correct that the passport-based api have options they can pass in too...)

Reply:
We can use something as follows:

```ts
@authenticate('basic', { /* some options for the strategy */})
```